soa/service/s3.cc

```diff
@@ -412,38 +432,28 @@ performSync() const
                 message += (string("body (") + to_string(responseBody.size())
                             + " bytes):\n" + responseBody + "\n");
             }
+            if (!xmlError.first.empty()) {
+                message += ("S3 REST error code: "
+                            + xmlError.first
+                            + "; message: "
+                            + xmlError.second
+                            +"\n");
+            }
 
+            return message;
+        };
 
-            /* log so-called "REST error"
-               (http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html)
-            */
-            if (responseHeaders.find("Content-Type: application/xml")
-                != string::npos) {
-                unique_ptr<tinyxml2::XMLDocument> localXml(
-                    new tinyxml2::XMLDocument()
-                );
-                localXml->Parse(responseBody.c_str());
-                auto element
-                    = tinyxml2::XMLHandle(*localXml).FirstChildElement("Error")
-                    .ToElement();
-                if (element) {
-                    message += ("S3 REST error: ["
-                                + extract<string>(element, "Code")
-                                + "] message ["
-                                + extract<string>(element, "Message")
-                                +"]\n");
-                }
-            }
+        if (!xmlError.first.empty()
+            || (responseCode >= 300 && responseCode != 404)) {
+            string message(makeErrorMsg());
             ::fprintf(stderr, "%s\n", message.c_str());
 
             /* retry on 50X range errors (recoverable) */
-            if (responseCode >= 500 && responseCode < 505) {
+            if ((responseCode >= 500 && responseCode < 505)
+                || xmlError.first == "InternalError") {
```

Comment:
This seems to be another weirdness with the S3 API.  We could get InternalError with a status code != 500.  If we have a link to this bug/feature we should add it here.